File: gem/lib/mulang/code.rb
module Mulang
  class Code
    attr_accessor :language, :content
    def initialize(language, content)
      @language = language
      @content  = content
    end

    def identifiers(**options)
      @language.identifiers @content, **options
    end

    def ast(**options)
      @language.ast @content, **options
    end

    def ast_analysis(**options)
      @language.ast_analysis @content, **options
    end

    def transformed_asts(operations, **options)
      @language.transformed_asts @content, operations, **options
    end

    def transformed_asts_analysis(operations, **options)
      @language.transformed_asts_analysis @content, operations, **options
    end

    def sample
      @language.sample @content
    end

    def analysis(spec)
      { sample: sample, spec: spec }
    end

    def analyse(spec, **options)
      Mulang.analyse analysis(spec), **options
    end

    def expect(binding='*', inspection)
      expectation = Mulang::Expectation.parse(binding: binding, inspection: inspection).as_v2.to_h
      expectation_results_for(analyse(expectations: [expectation])).first['result']
    end

    def custom_expect(edl)
      expectation_results_for(analyse(customExpectations: edl))
        .map { |e| [e['expectation']['inspection'], e['result']] }
        .to_h
    end

    alias query expect
    alias custom_query custom_expect

    def self.native(language_name, content)
      new Mulang::Language::Native.new(language_name), content
    end

    def self.native!(*args)
      native(*args).tap { |it| it.expect('Parses') }
    end

    def self.external(language_name = nil, content, &tool)
      new Mulang::Language::External.new(language_name, &tool), content
    end

    def self.analyse_many(codes, spec, **options)
      run_many(codes, **options) { |it| it.analysis(spec)  }
    end

    def self.ast_many(codes, **options)
      run_many(codes, key: 'outputAst', **options) { |it| it.ast_analysis(**options) }
    end

    def self.transformed_asts_many(codes, operations, **options)
      run_many(codes, key: 'transformedAsts', **options) { |it| it.transformed_asts_analysis(operations, **options) }
    end

    private

    def self.run_many(codes, key: nil, **options)
      result = Mulang.analyse(codes.map { |it| yield it }, **options)
      key ? result.map { |it| it[key] } : result
    end

    def expectation_results_for(result)
      raise result['reason'] if result['tag'] == 'AnalysisFailed'
      result['expectationResults']
    end
  end
end

File: ghcjslib/src/code.js
(() => {
  class MulangNativeLanguage {
    constructor(language) {
      this.language = language;
    }

    sample(content) {
      return { tag: 'CodeSample', language: this.language, content };
    }

    ast(code) {
      return code.computedAst;
    }
  }

  class MulangExternalLanguage {
    constructor() {
    }

    sample(content) {
      return { tag: 'MulangSample', ast: content };
    }

    ast(code) {
      return code.content;
    }
  }

  class MulangCode {
    constructor(language, content) {
      this.language = language;
      this.content = content;
    }

    // Returns the corresponding sample element
    // that is expected by Mulang's Analysis
    get sample() {
      return this.language.sample(this.content);
    }

    // Returns the ast for the given code
    // if the code is native, this ast is first computed by mulang, but
    // if this code is external, the original ast is returned instead
    get ast() {
      return this.language.ast(this);
    }

    get computedAst() {
      return this.analyse({expectations: [], smellsSet: { tag: 'NoSmells'}, includeIntermediateLanguage: true}).intermediateLanguage;
    }

    expect(binding, inspection) {
      return this
              .analyse({expectations: [{binding, inspection}]})
              .expectationResults[0]
              .result;
    }

    customExpect(edl) {
      return Object
        .values(this.analyse({customExpectations: edl}).expectationResults)
        .map((e) => [e.expectation.inspection, e.result]);
    }

    // Runs a mulang analysis, given an spec
    // Throws an error if analysis fails
    analyse(spec) {
      const result = mulang.analyse({ sample: this.sample, spec: spec });
      if (result.tag == 'AnalysisFailed') {
        throw new Error(result.reason);
      }
      return result;
    }
  }

  ghcjsExports.MulangCode = MulangCode;
  ghcjsExports.MulangNativeLanguage = MulangNativeLanguage;
  ghcjsExports.MulangExternalLanguage = MulangExternalLanguage;
})();
